### PR TITLE
[FIX] use first level object as input for second level

### LINKED ISSUE #3579: Problem with flm_objects as input to non_parametric_inference()
The non_parametric_inference() function uses the _check_second_level_input() function whose "flm_object" argument is set to False. This raises an error when trying to pass first-level model objects as second level input to the function. The function takes no "flm_object" arguments, so there is no way to change this parameter to True, thus allowing first-level model objects. 

Nilearn version: 0.10.0

### Expected behavior
If the "flm_object" argument is necessary in the _check_second_level_input() function, the non_parametetric_inference() function should take an argument that allows the user to change this parameter from False to True. 

### Actual behavior
It is not possible to change the "flm_object" argument from False to True, and thus not possible to pass first-level model objects to the second_level_input argument. 

### Steps and code to reproduce bug

1) Import modules 

```python
from nilearn.datasets import fetch_language_localizer_demo_dataset
from nilearn.glm.first_level import first_level_from_bids
from nilearn.glm.second_level import non_parametric_inference
```

2) Extract first-level model objects from BIDS data 

```python
data_dir, _ = fetch_language_localizer_demo_dataset()
task_label = 'languagelocalizer'
models, models_run_imgs, models_events, models_confounds = \
    first_level_from_bids(
        data_dir, task_label,
        img_filters=[('desc', 'preproc')])
```

3) Fit first-level models 

```python
model_and_args = zip(models, models_run_imgs, models_events, models_confounds)
for m_idx, (model, imgs, events, confounds) in enumerate(model_and_args):
     model.fit(imgs, events, confounds)
     model.compute_contrast('language-string')
```

4) Run non-parametric permutation using first-level models as second-level input 

```python
design_matrix = pd.DataFrame(
    [1] * len(second_level_input),
    columns=['intercept'],
)

n_perm=1000

out_dict = non_parametric_inference(
    second_level_input=models,
    design_matrix=design_matrix,
    first_level_contrast='language-string',
    model_intercept=True,
    n_perm=n_perm,  
    two_sided_test=False,
    smoothing_fwhm=8.0,
    n_jobs=1,
)
```

### PR DESCRIPTION
**Pull request by Remi-Gau**
Closes #3579

Changes proposed in this pull request:

- [x] add test to show bug from #3579
- [x] infer how to validate arguments from argument type in `non_parametric_inference`
- [x] update doc string to mention that a list of flm objects can be passed

### CODE REVIEW

File: nilearn/glm/second_level/second_level.py [hunk 1]
```diff
@@ -606,6 +606,7 @@ def non_parametric_inference(
     verbose=0,
     threshold=None,
     tfce=False,
+    flm_object=False
```
Comment (Remi-Gau) [inline nilearn/glm/second_level/second_level.py:609]: Although this fixes the test, I am not a big fan of this approach.

Would be better to not add extra arguments to the API.

File: nilearn/glm/second_level/second_level.py [hunk 2]
```diff
@@ -30,79 +30,82 @@
...
-def _check_second_level_input(second_level_input, design_matrix,
-                              confounds=None, flm_object=True, df_object=True):
+def _check_second_level_input(second_level_input,
+                              design_matrix,
+                              confounds=None):
     """Checking second_level_input type"""
     # Check parameters
     # check first level input
-    if isinstance(second_level_input, list):
+
```
Comment (NicolasGensollen) [inline nilearn/glm/second_level/second_level.py:39]: I find this function long and difficult to understand. I would break the different pieces into (almost) single responsibility functions. Maybe something more or less like this:

```python
def _check_second_level_input(second_level_input, design_matrix, confounds=None):
    input_type = _check_input_type(second_level_input)
    _check_input_as_type(
        second_level_input, input_type, confounds is None, design_matrix is None
    )

def _check_input_type(second_level_input) -> str:
    """Determines the type of input provided."""
    if isinstance(second_level_input, pd.DataFrame):
         return "df_object"
    if isinstance(second_level_input, (str, Nifti1Image)):
         return "nii_object"
    if isinstance(second_level_input, list):
        return _check_input_type_when_list(second_level_input)
    raise TypeError("....")

def _check_input_type_when_list(second_level_input: listl) -> str:
    """Determines the type of input provided when it is a list."""
    if len(second_level_input) < 2:
        raise TypeError("...")
    _check_all_elements_of_same_type(second_level_input)   
    if all(isinstance(x, (str, Nifti1Image)) for x in second_level_input):
        return "nii_object"
    if all(isinstance(x, FirstLevelModel) for x in second_level_input):
        return "flm_object"
    raise TypeError("...")

def _check_all_elements_of_same_type(data: list) -> None:
    for idx, input in enumerate(data):
        if not isinstance(input, type(data[0])):  
             raise TypeError(
                 f'Elements of second_level_input must be of the same type.'
                 f' Got object type {type(input)} at idx {idx}.')

def _check_input_as_type(second_level_input, input_type, none_confounds: bool, none_design_matrix: bool):
    if input_type == "flm_object":
        _check_input_as_first_level_model(second_level_input, none_confounds)
    elif input_type == "nii_object":
         _check_input_as_nifti_images(second_level_input, none_design_matrix)
    else:
         _check_input_as_dataframe(second_level_input)

def _check_input_as_first_level_model(second_level_input, none_confounds: bool):
    for model_idx, first_level in enumerate(second_level_input):
        if (first_level.labels_ is None or first_level.results_ is None):
             raise ValueError(
                 'Model %s at index %i has not been fit yet'
                 '' % (first_level.subject_label, model_idx))
         if not none_confounds and first_level.subject_label is None:
             raise ValueError("...")

def _check_input_as_dataframe(second_level_input):
    for col in ('subject_label', 'map_name', 'effects_map_path'):
        if col not in second_level_input.columns:
            raise ValueError("...")
    if not all(isinstance(_, str) for _ in second_level_input['subject_label'].tolist()):
        raise ValueError('...')

def _check_input_as_nifti_images(second_level_input, none_design_matrix):
    if isinstance(second_level_input, (str, Nifti1Image)):
        second_level_input = [second_level_input]
    for niimg in second_level_input:
        check_niimg(niimg=niimg, ensure_ndim=4)
    if none_design_matrix:
         raise ValueError("...")
```

**Review by NicolasGensollen — COMMENTED** [review]
Thanks @Remi-Gau !

I think this is looking good. When reviewing the second level input checking code, I realized that the associated function is really long and not super easy to understand. I think it could be the opportunity to refactor it a bit. I made a suggestion for what such a refactoring could look like. Let me know what you think.

Small question: Was the file `mask.nii` intended to be committed ?

**Comment by Remi-Gau** [comment]
> Small question: Was the file `mask.nii` intended to be committed ?

Nope.

That's a side effect of `test_infer_effect_maps()` that does not use a temp_dir.

I tried to play with it in another PR but could not fix it.
https://github.com/nilearn/nilearn/pull/3525/files#r1121367927

Will try to use pytest tmp_path instead of the nibabel TempDir thingy to if I can fix it.

File: nilearn/glm/second_level/second_level.py (hunk 2 above)
Comment (Remi-Gau) [reply nilearn/glm/second_level/second_level.py:39]: I did have it flagged as one of the function that was too complex and needed refactoring so I am totally in.

Let me incorporate that into the PR.

File: nilearn/glm/tests/test_second_level.py [hunk 3]
```diff
@@ -24,12 +24,14 @@
...
 from nilearn.glm.second_level import (SecondLevelModel,
                                       non_parametric_inference)
+from nilearn.reporting import get_clusters_table
 from scipy import stats
 
 # This directory path
 BASEDIR = os.path.dirname(os.path.abspath(__file__))
 FUNCFILE = os.path.join(BASEDIR, 'functional.nii.gz')
 
+N_PERM = 10
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_second_level.py:34]: use fewer permutation to speed up testing

**Comment by Remi-Gau** [comment]
Any idea why the CI failure "python 3.7 without matplotlib" is happening? 

Cannot see why this would be the case?

Apart from moving a `from nilearn.reporting import get_clusters_table` from within a test function, I don't have a hunch on why thus may be the case...

**Comment by ymzayek** [comment]
@Remi-Gau `get_clusters_table` relies on matplotlib. You can skip that test if matplotlib is not installed by using the `pytest.mark.skipif` decorator. See an an example the first test in `reporint.tests.test_reporting.py`

File: doc/changes/latest.rst [hunk 4]
```diff
@@ -22,8 +22,11 @@ Fixes
 
 - :func:`~nilearn.interfaces.fmriprep.load_confounds` can support searching preprocessed data in native space. (:gh:`3531` by `Hao-Ting Wang`_)
 
+- :func:`~nilearn.second_level.non_parametric_inference` can accept fist level model as input without failing. (:gh:`3600` by `Rémi Gau`_)
```
Comment (ymzayek) [inline doc/changes/latest.rst:25]:
```suggestion
- :func:`~nilearn.second_level.non_parametric_inference` can accept first level model as input without failing. (:gh:`3600` by `Rémi Gau`_)
```

**Review by ymzayek — APPROVED** [review]
LGTM thanks @Remi-Gau !
There are some type hints thrown around, probably done before our meeting, so maybe remove those? They are in private functions so it's not the biggest deal (and not sure if it's worth undoing work but also want to respect group decisions)

**Comment by Remi-Gau** [comment]
> There are some type hints thrown around, probably done before our meeting, so maybe remove those?

yeah better remove them for now and cross that bridge when we get to it

**Review by bthirion — APPROVED** [review]
LGTM.

**Review by NicolasGensollen — APPROVED** [review]
LGTM, thanks @Remi-Gau !

File: nilearn/glm/tests/test_second_level.py [hunk 5]
```diff
@@ -132,19 +172,13 @@ def test_check_second_level_input():
     with pytest.raises(ValueError,
                        match="List of niimgs as second_level_input "
                              "require a design matrix to be provided"):
-        _check_second_level_input("foo", None)
-    with pytest.raises(ValueError,
-                       match="second_level_input must be a list of"):
+        _check_second_level_input(fmri_data[0], None)
+    with pytest.raises(TypeError,
+                       match="second_level_input must be"):
+        _check_second_level_input(1, None)
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_second_level.py:178]: duplicate test can be removed
